### DISCUSSION TITLE
Using the library output in GLSL mesh shader

### DISCUSSION THREAD

**Discussion by Yossi4x6fu** [zeux/meshoptimizer#744] [Q&A] [answered]
Hello all,

I am quite confused about how to use the output from the library in a GLSL mesh shader.
From what I understand, 'meshopt_buildMeshlets' returns a vector of 'meshopt_Meshlet' with four parameters, two offsets for the indices and counts for each of them. The offsets are offsets into the global vertices and indices of the model. 
Inside the GLSL mesh shader I need to fill 'gl_PrimitiveIndicesNV' with local indices. I can scan the global indices at the given offset and fill the vertices read from the global vertices inside a local vertices array. I am not sure what to do with the vertices offset. I probably miss something here. 
A short code sample of how to use the meshlet inside a mesh shader will help a lot, or at least a short explanation of how to do it.
Many thanks,
Yossi

**Comment by zeux** [comment, marked as answer]
Let me know if you have ideas for how to improve documentation for meshopt_Meshlet structure (https://github.com/zeux/meshoptimizer/blob/master/src/meshoptimizer.h#L473-L489) or mesh shading section (https://github.com/zeux/meshoptimizer/blob/master/README.md#mesh-shading) except for adding a GLSL example - which I will do.

The vertices offset refers to the offset in the vertex indirection buffer (meshlet_vertices) and is just there so that from a mesh shader you can quickly locate the relevant section. Here's an example from https://github.com/zeux/niagara (reworked a little to simplify, untested but you can refer to the actual shader here https://github.com/zeux/niagara/blob/master/src/shaders/meshlet.mesh.glsl although be warned that it's more complex as it repacks the data a little for efficiency and uses large shared buffers for all meshes in a scene):

```glsl
layout(binding = 0) readonly buffer MeshletVertices
{
	uint meshletVertices[];
};

layout(binding = 1) readonly buffer MeshletIndices
{
	uint8_t meshletIndices[];
};
...

uint vertexCount = uint(meshlets[mi].vertexCount);
uint triangleCount = uint(meshlets[mi].triangleCount);

SetMeshOutputsEXT(vertexCount, triangleCount);

uint vertexOffset = uint(meshlets[mi].vertexOffset);
uint indexOffset = uint(meshlets[mi].indexOffset);

for (uint i = gl_LocalInvocationIndex; i < vertexCount; i += gl_WorkGroupSize.x)
{
  uint vi = meshletVertices[vertexOffset + i];

  vec3 position = vec3(vertices[vi].vx, vertices[vi].vy, vertices[vi].vz);

  gl_MeshVerticesEXT[i].gl_Position = globals.worldViewProjection * vec4(position, 1);
}

for (uint i = gl_LocalInvocationIndex; i < triangleCount; i += gl_WorkGroupSize.x)
{
  uint offset = indexOffset + i * 3;
  uint a = uint(meshletIndices[offset]), b = uint(meshletIndices[offset + 1]), c = uint(meshletIndices[offset + 2]);

  gl_PrimitiveTriangleIndicesEXT[i] = uvec3(a, b, c);
}
```

**Reply by Yossi4x6fu** [reply]
Many thank zeus for the fast and elaborated reply. Your answer helped me a lot!
I have still another question. How am trying to work an interleaved array of vertex/normal without any success.
I guess I should set the 'vertex_position_stride' when calling to 'meshopt_buildMeshlets' API to double its actual value (=24) but I am not sure it's the only setup needed.
I would much appreciate your help.

**Reply by zeux** [reply]
Yes, should just be a matter of changing the stride, [sample code](https://github.com/zeux/meshoptimizer/blob/master/demo/main.cpp#L908) uses interleaved vertices as well.

**Reply by Yossi4x6fu** [reply]
Many thanks zeux for the reply. Again, it helped me a lot and I can optimize now mesh with interleaved vertex/normal array.
I am trying now to compute the bounding parameters of the meshlets using 'meshopt_computeMeshletBounds' API.
The first parameter I suppose, is a pointer to the vertices indices returned by 'meshopt_buildMeshlets' which hold indices ids pointing to the vertices positions.
But when I call 'meshopt_computeMeshletBounds' with the first parameter set to the same vertex pointer (with the vertex offset of the meshlet), I get an assertion failure in line 878 file 'cluserizer.cpp':
assert(index < vertex_count)
The index is 681 and the vertex_count = 64. 
I probably misinterpret the meaning of the input parameters.
I would appreciate your help.
Many thanks

**Comment by Yossi4x6fu** [comment]
Many thanks zeux for the reply. Again, it helped a lot and I can optimize now mesh with interleaved vertex/normal array.
I am trying now to compute the bounding parameters of the meshlets using 'meshopt_computeMeshletBounds' API.
The first parameter I suppose, is a pointer to the vertices indices returned by 'meshopt_buildMeshlets' which hold indices ids pointing to the vertices positions.
But when I call 'meshopt_computeMeshletBounds' I get an assertion failure in line 878 file 'cluserizer.cpp':
assert(index < vertex_count)
The index is 681 and the vertex_count = 64. 
I probably misinterpret the meaning of the input parameters.
I would appreciate your help.
Many thanks

**Reply by zeux** [reply]
The usage is documented here https://github.com/zeux/meshoptimizer#mesh-shading; you are probably passing the vertex count of an individual meahlet, not the full mesh.
